src/test/ui/async-await/issues/issue-62009-1.stderr

```diff
@@ -24,15 +24,14 @@ LL | fn main() {
 LL |     (|_| 2333).await;
    |               ^^^^^^ only allowed inside `async` functions and blocks
 
-error[E0277]: `[closure@$DIR/issue-62009-1.rs:12:6: 12:9]` is not a future
+error[E0277]: `[closure@$DIR/issue-62009-1.rs:12:6: 12:9]` cannot be converted to a future
   --> $DIR/issue-62009-1.rs:12:15
    |
 LL |     (|_| 2333).await;
-   |               ^^^^^^ `[closure@$DIR/issue-62009-1.rs:12:6: 12:9]` is not a future
+   |               ^^^^^^ `[closure@$DIR/issue-62009-1.rs:12:6: 12:9]` cannot be converted to a future
```

Comment:
I really hate how we display closures 😩

Ideally we should number them and display a span label naming all the ones that are involved, but this will require handling that on every error independently :-/

Reply:
[related discussion on irlo](https://internals.rust-lang.org/t/extending-implicit-coercion-of-function-items-to-function-pointers-during-trait-resolution/17093/6?u=cad97) wants to display fn items more like closures, specifically so they don't look like fn pointers.

Perhaps lying and calling the closure `impl FnOnce(_) -> {integer}` would work? Or not abusing impl trait, `{closure}: FnOnce(_) -> {integer}`.

There is a potential way to number closures automatically, but it would still require touching every location that's naming types; roughly one of:

- `fn DiagnosticBuilder::name_type(ty: &Ty) -> impl Display`; provides an indexed name for `{closure#0}` and registers to show a map of indexed to the normal display formatting.
- Add functionality to `Print`/`PrettyPrinter` to do this numbering, *require* printing types to use `Print` instead of `Display` (e.g. here using `DiagnosticBuilder` as the printing context), and do the `name_type` stuff through the existing `Print`/`PrettyPrinter` machinery.

Either way it seems reasonable to add a `PrettyPrinter::pretty_print_closure` or similar.